File: pisa-daemon/pkg/tc/tc.go
// Copyright 2022 SphereEx Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package tc

import (
	// "C"
	"sort"
	"strings"

	v1alpha1 "github.com/database-mesh/waterline/api/v1alpha1"
	"github.com/mlycore/log"
	"github.com/vishvananda/netlink"
	"golang.org/x/sys/unix"
	"k8s.io/apimachinery/pkg/api/resource"
)

type Shaper struct {
	qos            v1alpha1.TrafficQoS
	link           netlink.Link
	totalBandWidth string
}

func NewTcShaper(qos v1alpha1.TrafficQoS, totalBandWidth string) (*Shaper, error) {
	link, err := netlink.LinkByName(qos.Spec.NetworkDevice)
	if err != nil {
		return nil, err
	}

	return &Shaper{
		qos:            qos,
		link:           link,
		totalBandWidth: totalBandWidth,
	}, nil
}

// add htb qidsc, called by AddClasses
func (t *Shaper) addHtbQdisc() error {
	attrs := netlink.QdiscAttrs{
		LinkIndex: t.link.Attrs().Index,
		Handle:    netlink.MakeHandle(1, 0),
		Parent:    netlink.HANDLE_ROOT,
	}

	qdisc := netlink.NewHtb(attrs)
	log.Infof("htb qdisc attrs: %#v, qdisc: %#v", attrs, qdisc)
	return netlink.QdiscReplace(qdisc)
}

// add htb root handle, called by AddClasses
func (t *Shaper) addRootHandle() error {
	attrs := netlink.ClassAttrs{
		LinkIndex: t.link.Attrs().Index,
		Parent:    netlink.MakeHandle(1, 0),
		Handle:    netlink.MakeHandle(1, 1),
	}

	bandValue, err := resource.ParseQuantity(t.totalBandWidth)
	if err != nil {
		return err
	}

	htbClassAttrs := netlink.HtbClassAttrs{
		Rate: uint64(bandValue.Value()),
	}

	class := netlink.NewHtbClass(attrs, htbClassAttrs)
	log.Infof("add root handle attrs: %#v, htbClassAttrs: %#v, qdisc: %#v", attrs, htbClassAttrs, class)
	return netlink.ClassReplace(class)
}

func (t *Shaper) AddClasses() ([]int, error) {
	err := t.addHtbQdisc()
	if err != nil && !strings.Contains(err.Error(), "invalid argument") {
		log.Errorf("add htb qdisc error: %s", err)
		return err
	}

	err = t.addRootHandle()
	if err != nil && !strings.Contains(err.Error(), "invalid argument") {
		log.Errorf("add root handle error: %s", err)
		return err
	}

	rules := t.qos.Spec.Groups
	// sort by rate
	sort.Slice(rules, func(i, j int) bool {
		return rules[i].Rate < rules[j].Rate
	})

	classes, err := t.ListClass()
	if err != nil {
		log.Errorf("list class error: %s", err)
		return err
	}

	var base uint16
	for _, c := range classes {
		// c.Attrs().Handle
		_, minor := netlink.MajorMinor(c.Attrs().Handle)
		if base < minor {
			base = minor
		}
	}

	//TODO: add error handling.
	for idx, rule := range rules {
		if err := t.addClass(uint16(idx)+base, rule); err != nil {
			log.Errorf("add class error: %s, rule: %s", err, rule)
			return err
		}
	}

	return nil
}

// add htb class
func (t *Shaper) addClass(idx uint16, rule v1alpha1.TrafficQoSGroup) error {
	attrs := netlink.ClassAttrs{
		LinkIndex: t.link.Attrs().Index,
		Parent:    netlink.MakeHandle(1, 1),
		//exclude 0, 1
		Handle: netlink.MakeHandle(1, idx+1),
	}

	rateValue, err := resource.ParseQuantity(rule.Rate)
	if err != nil {
		return err
	}

	htbClassAttrs := netlink.HtbClassAttrs{
		Rate: uint64(rateValue.Value()),
	}

	switch t.qos.Spec.QoSClass {
	case v1alpha1.QoSClassGuaranteed:
		if rule.Ceil == "" {
			ceilValue, err := resource.ParseQuantity(rule.Rate)
			if err != nil {
				return err
			}
			htbClassAttrs.Ceil = uint64(ceilValue.Value())
		}
	case v1alpha1.QoSClassBurstable:
		if rule.Ceil == "" {
			htbClassAttrs.Ceil = uint64(float64(rateValue.Value()) * 1.5)
		}
	case v1alpha1.QoSClassBestEffort:
		return nil
	}

	class := netlink.NewHtbClass(attrs, htbClassAttrs)
	return netlink.ClassReplace(class)
}

// add qdisc clsact
func (t *Shaper) addClsAct() error {
	attrs := netlink.QdiscAttrs{
		LinkIndex: t.link.Attrs().Index,
		Handle:    netlink.MakeHandle(0xffff, 0),
		Parent:    netlink.HANDLE_CLSACT,
	}

	qdisc := netlink.GenericQdisc{
		QdiscAttrs: attrs,
		QdiscType:  "clsact",
	}

	return netlink.QdiscReplace(&qdisc)
}

// match qdisc func
type matchQdiscFunc = func(qdisc netlink.Qdisc) bool

// delete matched qdisc
func (t *Shaper) delQdisc(f matchQdiscFunc) error {
	qdiscs, err := netlink.QdiscList(t.link)

	if err != nil {
		return err
	}

	for _, v := range qdiscs {
		if f(v) {
			return netlink.QdiscDel(v)
		}
	}

	return nil
}

// match class func
type matchClassFunc = func(class netlink.Class) bool

// delete matched class
func (t *Shaper) delClass(f matchClassFunc) error {
	classes, err := t.ListClass()
	if err != nil {
		return err
	}

	for _, v := range classes {
		if f(v) {
			return netlink.ClassDel(v)
		}
	}

	return nil
}

// ListClass list class
func (t *Shaper) ListClass() ([]netlink.Class, error) {
	return netlink.ClassList(t.link, netlink.MakeHandle(1, 0))
}

// AddFilter add bpf filter, default obj name is "tc.o", reqired `addClsAct`
func (t *Shaper) AddFilter() error {
	filterAttrs := netlink.FilterAttrs{
		LinkIndex: t.link.Attrs().Index,
		Parent:    netlink.MakeHandle(1, 0),
		Protocol:  unix.ETH_P_ALL,
	}

	bpfFilter := netlink.BpfFilter{
		FilterAttrs:  filterAttrs,
		ClassId:      netlink.MakeHandle(1, 0),
		Name:         "tc.o",
		DirectAction: true,
	}

	return netlink.FilterAdd(&bpfFilter)
}

File: pisa-daemon/tc/src/tc.rs
// Copyright 2022 SphereEx Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::process::Command;

#[derive(Clone)]
pub struct QdiscRootAttr<'a> {
    pub netns: Option<&'a str>,
    pub device: &'a str,
    pub typ: &'a str,
}

pub fn add_root_qdisc<'a>(attr: &QdiscRootAttr<'a>) -> bool {
    let mut args = vec![
        "qdisc",
        "add",
        "dev",
        &attr.device,
        "root",
        "handle",
        "1:",
        &attr.typ,
    ];

    execute_tc_command!( attr, args, "add root qdisc")

}

pub fn delete_root_qdisc<'a>(device: &str) -> bool {
    let args = vec![
        "qdisc",
        "delete",
        "dev",
        device,
        "root",
        "handle",
        "1:",
    ];

    let out = Command::new("tc")
        .args(&args)
        .output()
        .expect("faild to delete root qdisc");


    if !out.status.success() {
        println!("Failed to delete root qdisc to device {}: {:?}", device, out);
        return false;
    }

    true
}

pub fn show_qdisc(device: String) -> bool {
    let out = Command::new("tc")
        .args(["qdisc", "show", "dev", &device])
        .output()
        .expect("faild to show qdisc");

    if !out.status.success() {
        println!("Failed to show qdisc to device {}: {:?}", device, out);

        return false;
    }

    true
}

pub struct ClassAttr<'a> {
    pub netns: Option<&'a str>,
    pub device: &'a str,
    pub parent: Option<&'a str>,
    pub class_id: &'a str,
    pub rate: &'a str,
    pub ceil: Option<&'a str>,
}

pub fn add_class<'a>(attr: &ClassAttr<'a>) -> bool {
    let mut args = vec![
        "class",
        "add",
        "dev",
        &attr.device,
        "parent",
        &attr.parent.unwrap_or("1:"),
        "classid",
        &attr.class_id,
        "htb",
        "rate",
        &attr.rate,
    ];

    if let Some(ceil) = &attr.ceil {
        args.push("ceil");
        args.push(ceil);
    }


    execute_tc_command!( attr, args, "add class")

}

pub fn delete_class<'a>(attr: &ClassAttr<'a>) -> bool {
    let mut args = vec![
        "class",
        "delete",
        "dev",
        &attr.device,
        "parent",
        "1:",
        "classid",
        &attr.class_id,
        "htb",
        "rate",
        &attr.rate,
    ];


    if let Some(ceil) = &attr.ceil {
        args.push("ceil");
        args.push(ceil);
    }


    execute_tc_command!( attr, args, "delete class")

}

pub fn show_class(netns: Option<&str>, device: &str) -> (Vec<ClassResult>, bool) {
    let mut args = vec!["class", "show", "dev", &device];

    if let Some(ns) = netns {
        args.insert(0, "-n");
        args.insert(1, ns);
    }
    
    let out = Command::new("tc")
        .args(&args)
        .output()
        .expect("faild to show class");

    if !out.status.success() {
        println!("Failed to show class to device {}: {:?}", device, out);

        return (vec![], false);
    }

   let lines  = std::str::from_utf8(&out.stdout).unwrap().lines();

   let mut res = vec![];
   for line in lines {
        let arr = line.split(" ").map(|x| x.to_string()).collect::<Vec<_>>();
       res.push(ClassResult::from(arr));

   }

    (res, true)
}

#[derive(Debug)]
pub enum ClassResult {
    Htb(Htb)
}

#[derive(Debug)]
pub struct Htb {
    pub parent: String,
    pub class_id: String,
    pub rate: String,
    pub ceil: String,
}

impl  From<&[String]> for Htb {
    fn from(input: &[String]) -> Self {
        let parent = if &input[1] == "root" {
            "root"
        } else {
           &input[2]
        };

        let rate_idx = input.iter().position(|x| x == "rate").unwrap();
        let ceil_idx = input.iter().position(|x| x == "ceil").unwrap();

        Htb { 
            parent: parent.to_string(), 
            class_id: input[0].clone(), 
            rate: input[rate_idx+1].clone(), 
            ceil: input[ceil_idx+1].clone(),
        }
    }
}

impl From<Vec<String>> for ClassResult {
    fn from(input: Vec<String>) -> ClassResult {
        match input[1].as_str() {
           "htb" => {
                ClassResult::Htb(Htb::from(&input[2..]))
           },
           _ => unreachable!()
        }
    }
}


#[cfg(test)]
mod test {
    use super::*;
    
    fn create_test_ns(ns: &str) {
        let out = Command::new("ip")
            .args(["netns", "add", ns])
            .output()
            .unwrap();
        println!("creat test ns {:?}", out);
    }

    fn delete_test_ns(ns: &str) {
        let out = Command::new("ip")
            .args(["netns", "delete", ns])
            .output()
            .unwrap();
        println!("delete test ns {:?}", out);
    }


    fn create_root_qdisc(ns: &str) -> bool {
        let attr = QdiscRootAttr {
            netns: Some(ns),
            device: "lo",
            typ: "htb",
        };

        add_root_qdisc(&attr)
    }

    #[test]
    fn test_create_root_qdisc() {
        create_test_ns("footest1");
        assert_eq!(create_root_qdisc("footest1"), true);
        delete_test_ns("footest1");
    }

    #[test]
    fn test_create_delete_class() {
        create_test_ns("footest2");

        create_root_qdisc("footest2");

        let attr = ClassAttr {
            netns: Some("footest2"),
            device: "lo",
            class_id: "1:10",
            rate: "1mbps",
            ceil: None,
            parent: None,
        };

        assert_eq!(add_class(&attr), true);
        assert_eq!(delete_class(&attr), true);

        delete_test_ns("footest2");
    }

    #[test]
    fn test_show_class() {
        create_test_ns("footest");
        create_root_qdisc("footest");

        let attr = ClassAttr {
            netns: Some("footest"),
            device: "lo",
            class_id: "1:10",
            rate: "1mbps",
            ceil: None,
            parent: Some("1:1"),
        };

        add_class(&attr);

        let (res, _ ) = show_class(Some("footest"), "lo");

        let ClassResult::Htb(htb) = &res[0];
        {
            assert_eq!(htb.class_id, "1:10");
        }
        
        delete_test_ns("footest");
    }
}
